File: data/src/main/java/com/arturogutierrez/openticator/mapper/Mapper.java
package com.arturogutierrez.openticator.mapper;

import java.util.ArrayList;
import java.util.Collection;
import java.util.List;

public abstract class Mapper<FROM, TO> {

  public abstract TO transform(FROM values);

  public abstract FROM reverseTransform(TO values);

  public List<TO> transform(Collection<FROM> values) {
    List<TO> transformedValues = new ArrayList<>(values.size());
    for (FROM value : values) {
      transformedValues.add(transform(value));
    }
    return transformedValues;
  }

  public List<FROM> reverseTransform(Collection<TO> values) {
    List<FROM> transformedValues = new ArrayList<>(values.size());
    for (TO value : values) {
      transformedValues.add(reverseTransform(value));
    }
    return transformedValues;
  }
}

File: data/src/main/kotlin/com/arturogutierrez/openticator/mapper/Mapper.kt
package com.arturogutierrez.openticator.mapper

interface Mapper<FROM, TO> {

  fun transform(value: FROM): TO

  fun reverseTransform(value: TO): FROM

  fun transform(values: Collection<FROM>) = values.map { transform(it) }

  fun reverseTransform(values: Collection<TO>) = values.map { reverseTransform(it) }
}
